lib/sinatra/base.rb

```diff
@@ -294,13 +299,18 @@ def body(value = nil, &block)
         def block.each; yield(call) end
         response.body = block
       elsif value
-        unless request.head? || value.is_a?(Rack::Files::Iterator) || value.is_a?(Stream)
-          headers.delete 'content-length'
-        end
         response.body = value
-      else
-        response.body
+
+        # If this isn't a head request, try to recompute the content length:
+        if !request.head?
+          body = response.body
+          if body.is_a?(Array)
+            headers['content-length'] = body.sum(&:bytesize).to_s
+          end
+        end
```

Comment:
Wont this try to calculate the content length more often than what the current solution does?

https://github.com/sinatra/sinatra/blob/5640495babcb4cfd69ba650b293660b7446402da/lib/sinatra/base.rb#L193-L206